src/services/cube-builder.ts

```diff
@@ -1,861 +1,1199 @@
+  buildRevisionId: string,
+  buildType = CubeBuildType.FullCube,
+  buildId = crypto.randomUUID()
 ): Promise<void> => {
-  logger.debug(`Loading data table ${dataTableId} from data_tables schema into cube fact table`);
-  const insertQuery = pgformat(
-    'INSERT INTO %I SELECT %I FROM %I.%I;',
-    factTableName,
-    factTableDef,
-    'data_tables',
-    dataTableId
-  );
+  const datasetRelations: FindOptionsRelations<Dataset> = {
+    factTable: true,
+    dimensions: { metadata: true, lookupTable: true },
+    measure: { metadata: true, measureTable: true },
+    revisions: { dataTable: { dataTableDescriptions: true } }
+  };
+
+  logger.debug('Loading dataset and relations');
+  const dataset = await DatasetRepository.getById(datasetId, datasetRelations);
+  logger.debug('Loading revision and relations');
+  const buildRevision = dataset.revisions.find((rev) => rev.id === dataset.endRevisionId);
+
+  if (!buildRevision) {
+    logger.error('Unable to find buildRevision in dataset.');
+    throw new CubeValidationException('Failed to find buildRevision in dataset.');
+  }
+
+  const cubeBuildConfig = cubeConfig.map((config) => {
+    const columns = new Map<Locale, Set<string>>();
+    const viewParts = new Map<Locale, string[]>();
+    SUPPORTED_LOCALES.forEach((locale) => {
+      columns.set(locale, new Set<string>());
+      viewParts.set(locale, []);
+    });
+    return {
+      name: config.name,
+      config: config,
+      columns,
+      viewParts
+    } as CubeViewBuilder;
+  });
+  logger.debug(`Build type = ${buildType}`);
+
+  let build: BuildLog;
+  try {
+    build = await BuildLog.startBuild(buildRevision, buildType, buildId);
+  } catch (err) {
+    logger.error(err, 'Failed to create build log entry');
+    throw err;
+  }
+
+  const createBuildSchemaRunner = dbManager.getCubeDataSource().createQueryRunner();
   try {
-    await cubeDB.query(insertQuery);
+    logger.info(`Creating schema for cube ${build.id}`);
+    await createBuildSchemaRunner.query(pgformat(`CREATE SCHEMA IF NOT EXISTS %I;`, build.id));
   } catch (error) {
-    logger.error(error, `Failed to load file into table using query ${insertQuery}`);
-    throw new FactTableValidationException(
-      'An unknown error occurred trying to load data in to the fact table.  Please contact support.',
-      FactTableValidationExceptionType.UnknownError,
-      500
-    );
+    logger.error(error, 'Something went wrong trying to create the cube schema');
+    throw error;
+  } finally {
+    void createBuildSchemaRunner.release();
+  }
+
+  let cubeBuild: CubeBuilder;
+  try {
+    cubeBuild = await createBasePostgresCube(build, dataset, buildRevision, buildType, cubeBuildConfig);
+  } catch (err) {
+    logger.error(err, 'createAllCubeFiles: Something went wrong during the actual build process');
+    if (config.cube_builder.preserve_failed) {
+      throw err;
+    }
+    const unwindBuildRunner = dbManager.getCubeDataSource().createQueryRunner();
```

Comment:
Doesn't look like this gets released. Any reason not to just reuse the existing query runner `createBuildSchemaRunner` here? That's being released after the unwind anyways.

I think generally it's ok to reuse query runners inside a single function - they only become difficult to keep track of when passing around between functions.

Reply:
Fixed it... It should be `void unwindBuildRunner.release()`... It doesn't cause a problem trying to release the otherwise released connection because it already throws an error to handle.